openedx2zim/xblocks_extractor/base_xblock.py

```diff
@@ -18,7 +35,34 @@ def __init__(
         # make xblock output directory
         self.output_path.mkdir(parents=True, exist_ok=True)
 
-    def download(self, instance_connection):
+        self.watcher.total_count += 1
+
+    @classmethod
+    def too_many_failures(cls):
+        return cls.watcher.dl_count > cls.watcher_min_dl_count and (cls.watcher.success_count / cls.watcher.dl_count) < cls.watcher_min_ratio
+    
+    def download(self, instance_connection, lock):
+        if BaseXblock.too_many_failures():
+            return
+        self.lock = lock
+        with self.lock:
+            self.watcher.dl_count += 1
+        logger.debug(f"Downloading resource {self.watcher.dl_count} of {self.watcher.total_count} ({self.watcher.success_count} success so far)")
+        try:
+            self.download_inner(instance_connection=instance_connection)
+        except Exception:
+            self.add_failed({})
+            return
+        with self.lock:
+            self.watcher.success_count += 1
+
+    def add_failed(self, description):
+        with self.lock:
+            description["xblock_id"] = self.xblock_id
+            description["class"] = type(self).__name__
+            self.watcher.failed_xblocks.append(description)
+
+    def download_inner(self, instance_connection):
```

Comment:
@rgaudin 
I don't get it sorry. Which `if content` would you like to save ?
If you speak about `if not content:`, I still think we need this because we have to take into account the fact that `get_page` might return either None or raise an Exception.

Reply:
Yes, I haven't looked at `get_page()` (hence the “I think”) but should it not return `None`, that what the idea

Reply:
OK, I get the idea. `get_page` is currently returning None when it failed, but I agree this is not correct (and will change in the other PR). 
We still need to `return` in `download_inner` because this is what allows us to not have to implement it in all xblocks, because some only need a rendering phase, and no download. But this is still ok.